IdentityCore/tests/automation/MSIDClientCredentialHelper.m

```diff
@@ -0,0 +1,249 @@
+// Copyright (c) Microsoft Corporation.
+// All rights reserved.
+//
+// This code is licensed under the MIT License.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files(the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDClientCredentialHelper.h"
+#import "NSData+MSIDExtensions.h"
+#import "MSIDJWTHelper.h"
+#import "MSIDLegacyTokenCacheKey.h"
+#import "MSIDAccessToken.h"
+#import "MSIDAADV1TokenResponse.h"
+#import "MSIDAADV1Oauth2Strategy.h"
+#import "NSDictionary+MSIDExtensions.h"
+
+@implementation MSIDClientCredentialHelper
+
++ (NSMutableDictionary<MSIDLegacyTokenCacheKey *, MSIDAccessToken *> *)accessTokenCache
+{
+    static dispatch_once_t once;
+    static NSMutableDictionary<MSIDLegacyTokenCacheKey *, MSIDAccessToken *> *accessTokenCache = nil;
+
+    dispatch_once(&once, ^{
+        accessTokenCache = [NSMutableDictionary dictionary];
+    });
+
+    return accessTokenCache;
+}
+
++ (void)getAccessTokenForAuthority:(NSString *)authority
+                          resource:(NSString *)resource
+                          clientId:(NSString *)clientId
+                       certificate:(NSData *)certificateData
+               certificatePassword:(NSString *)password
+                 completionHandler:(void (^)(NSString *accessToken, NSError *error))completionHandler
+{
+    MSIDLegacyTokenCacheKey *cacheKey = [MSIDLegacyTokenCacheKey keyWithAuthority:[NSURL URLWithString:authority]
```

Comment:
Q: Couldn't we check completionHandler quick return if there is no completionHandler?
Doesn't seem like we save the token back in this method, so I don't see a need to continue on if there is no completion handler here.

Reply:
It does cache the access token when it gets it back